doc/adr/0004-pab-indexing-solution-integration.rst

```diff
@@ -0,0 +1,183 @@
+
+Add indexing specific functions in the Contract API
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this scenario, we would need to split Contract API requests which interact with an external indexing component to the ones that use the PAB.
+Currently, we have `awaitTxConfirmed` which uses the indexed information in the PAB to wait for a transaction status to change to `Confirmed`.
+On top of that, we can have `awaitTxIndexed` or `awaitTxOutIndexed` which will wait for the information to be indexed in the external indexing component.
+
+Pros:
+
+- Limits design change on the PAB
+- More control given to the user of the Contract API
+
+Cons:
+
+- Adds an undesired complexity to the Contract API
+- We'll need to add a bunch of functions (e.g., `currentNodeSlot`, `currentMarconiSlot`, `awaitMarconiTxConfirmed, `awaitScrollsTxConfirmed`, etc.) for each new indexing component we want to support
+
+Query functions should interact with a single source of truth
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this scenario, we make the design decision that the Contract API should only interact with a single indexing component.
+Thus, any blockchain information currently stored in the PAB should be moved to the indexing component.
+Also, combining indexing components would need to be integrated in the single indexing solution that’s connected to the PAB.
+
+Pros:
+
+- Simplest in design to implement (other than manual work to move code)
+- No modification to the Contract API
+- Augments PAB's cohesion, because it's responsability will be limited to contract instance management
+
+Cons:
+
+- PAB won't be able to integrate with external indexing components (e.g. Blockfrost_ or Scrolls_), but these external solutions would need to be integrated into the single source of truth indexing solution. Therefore, this solution reduces the PAB's flexibility.
+- The design of the indexing component will need to be changed to support waiting queries (like the `awaitTxConfirmed` from PAB)
+
+Alter indexing query APIs to include valid slots information
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In this scenario, we can instead embrace the fact that the indexing component is not always in sync with the PAB.
+We can alter the indexing component query APIs (like `utxosAt` or `txOutFromRef`) to include some sort of `ValidSlotRange` information.
+Therefore, the user would need to specify the `SlotRange` from which he expects the indexing component to be synced.
+If the currently indexed slot of the PAB and indexing component is not up to what we need, we can either wait for it or throw an error.
+
+Pros:
+
+- Would work accross different indexers as long as we can get an indication as to the slot number they are synced up to (which is not unreasonable to assume)
+- Limits the changes to the Contract API
+
+Cons:
+
+- Makes the query APIs more complicated, which ultimately result in a more complex contract
```

Comment:
I am not sure this is the case. The contract API consists of a EDSL (effect system based), so we could implement synchronisation as a cross-cutting/semantic/interpreter concern, by specifying a 'context configuration' where the user would decide what is acceptable for their application in general.

Reply:
True! But the downside of this is that is doesn't actually solve the problem in the context right? The contract would throw a runtime error if the chain-index is not fully synced. Compare that to the "single source of truth" solution were you would never get a synchronization error.

The upside is that the error would be very clear and explicit.

Reply:
As @andreabedini noticed before, it's all about the implementation of the interpreter. You can have an interpreter that waits until all indexers have caught-up with the chain in which case you would never get a synchronisation error either.